### DISCUSSION TITLE
How to use Firedrake's external operator to calculate layer potential? help!!!!!!!!!!!!!!!!!!

### DISCUSSION THREAD

**Discussion by LMathtam** [firedrakeproject/firedrake#3650] [General]
During the calculation process, the boundary condition is given by an integral, and a layer potential needs to be calculated. How can I use the external operator of Firedrake to calculate it?

**Comment by colinjcotter** [comment]
I'm lacking context for your request. Please can you define what equations and boundary conditions you are solving, and what you want to calculate? Then we can decide if external operator is a good idea or something else.

**Reply by LMathtam** [reply]
Thanks for your reply. In fact, we can consider the Helmholtz equation on a rectangular region \Omega with the boundary conditions given by the single layer potential which looks like \int_{\partial \Omega} f(x)G(x,y)dx ,where f is a given function and G(x,y) is the fundamental solution of the 3d Helmholtz function.

**Reply by colinjcotter** [reply]
Thanks! Information on how to use ExternalOperator is here - you just need to provide the methods described. 
https://www.firedrakeproject.org/external_operators.html#id7
What I don't know is whether ExternalOperator can be used in a DirichletBC, hopefully someone like @nbouziani can answer!

**Reply by nbouziani** [reply]
Given that `ExternalOperator`s can be used in UFL expressions, I don't see any obstacles in using them for boundary conditions. That should just work. As for using it for layer potentials, @rckirby has been using external operators for layer potentials for a while now, so he can provide more insights on that particular application.

**Reply by rckirby** [reply]
I don't know how (mathematically) one would use it in a strong boundary condition.  We have a boundary condition appearing in the variational form:

https://epubs.siam.org/doi/pdf/10.1137/20M1368100

https://www.sciencedirect.com/science/article/pii/S0898122124000919

This a different method than Johnson/Nedelece FEM-BEM or Givoli-Keller DtN (no singular integrals).  I'm not sure what the current status of our code is vis-a-vis external operators landing, but it's at least a go-by:

https://github.com/inducer/nufl

We use Andreas Klöckner's pytential code stack to evaluate the layer potential via an FMM, and don't actually assemble the matrix (dense blocks!) but use a matrix-free technique.

**Reply by LMathtam** [reply]
Thanks for your nice reply!

**Reply by lrtfm** [reply]
@rckirby It seems the following link has expired. Could you please provide an updated one? Thank you!
> https://github.com/inducer/nufl
